### ISSUE TITLE
S3 Input Stream closes before file is done downloading

### ISSUE THREAD

**Issue by cin** [aws/aws-sdk-java#472]
I'm trying to stream a large, gzip file from S3 and am encountering a few issues. More often than not, the download stops about 1/4 of the way through (at exactly the same place). Sometimes the download stops about 1/2 way through (at the same place). Rarely, it completes. Part of the problem here is there is no error or anything generated. Unless you look at the number of lines processed, it appears as if the entire file was processed. The file in question is ~4GB.

Here's the basics of the code. It's really simple, but I'm obviously missing something (or there's a bug).

    val gzis = new GZIPInputStream(client.getObject(bucket, fn).getObjectContent)
    val br = new BufferedReader(new InputStreamReader(gzis, encoding))
    val (_, t) = time(processStream(br))
    br.close()
    gzis.close()

The processStream method is simply calling readLine on the BufferedReader until it returns null. I have spent some time trying to figure out what's going on but haven't gotten anywhere. Any assistance/insights would be appreciated.

Java 1.8.0_31
Scala 2.11.6
aws-java-sdk-s3 - 1.10.6 (was using 1.9.40 with same symptoms)

**Comment by manikandanrs** [comment]
Interesting. Have you tried without wrapping the stream from S3 with GZipInputStream.  There is a know bug with the GZipInputStream that returns 0 even when there is more data available in the underlying stream.

http://bugs.java.com/bugdatabase/view_bug.do?bug_id=7036144

**Comment by manikandanrs** [comment]
Also, one more question about the object that you are trying to download. Is it GZIP encoded ?

**Comment by cin** [comment]
I haven't tried removing the GZIPInputStream call. I'll try that and see if I can at least read all the bytes (even if I can't process them).

The content type is actually set to "binary/octet-stream". Shouldn't this be "application/gzip"? I actually don't control the files in S3 or their content, but it would be nice to understand why this is happening (if that's it).

The file is gzip encoded, and if I download the file I can process the entire file cleanly every time.

**Comment by manikandanrs** [comment]
The content-type for an object is set during a PUT Object call. By default it is set to binary/octet-stream.  Here is the API spec for Put Object
http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html

You can change the content type in the PutObject request by updating the object metadata in PutObjectRequest

https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/ObjectMetadata.java#L401

Let us know if you are able to download the object without wrapping the response stream with GZipInputStream. 

When a GZip encoded object is downloaded by S3, the SDK currently handles the bug in JDK.

https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/S3ObjectInputStream.java#L115

**Comment by cin** [comment]
So I heeded your advice and tried not using GZIPInputStream and just processed the raw, compressed bytes to see if I could get the number of bytes downloaded to match the size of the file. I was able to get that working every time. I then added the GZIPInputStream back into the mix but used a BufferedInputStream instead of a BufferedReader. This worked every time but was, of course, slow. Since this was working, it lead me to believe it was a configuration issue with how I was setting up the streams. I started calling constructors to GZIPInputStream and BufferedReader that take a buffer size and viola! I ended up bumping the buffer size up to 4096 and haven't experienced any more abnormalities.

It makes me really uncomfortable that I am depending on a tuning setting to make this reliable, but it's the best solutions I've come up with so far.

Looking at the available method in S3ObjectInputStream is concerning as well. Simply returning 1 when the parent returns a 0 seems to be more of a kludge than a real solution. It seems that whenever the client exhausts its local buffer, it's considering the stream done (pure guess). At least, that's the only way I can justify that increasing the buffer size fixed the problem.
